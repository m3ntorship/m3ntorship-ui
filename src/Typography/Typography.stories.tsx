import { Meta, Story } from '@storybook/react';
import React from 'react';

import { Typography } from './Typography';
import { TypographyProps } from './Typography.types';

export default {
  title: 'Typography',
  component: Typography,
} as Meta<typeof Typography>;

const Template: Story<TypographyProps> = args => <Typography {...args} />;

export const Default = Template.bind({});
Default.args = {
  variant: 'h1',
  children: 'Lorem ipsum dolor sit amet.',
  className: 'bg-red-200',
};
