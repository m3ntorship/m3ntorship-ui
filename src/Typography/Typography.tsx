import classNames from 'classnames';
import React, { ElementType, FC } from 'react';

import {
  BaseTypographyProps,
  SHARED_CLASSNAMES,
  TYPOGRAPHY_VARIANTS,
  TypographyProps,
} from './Typography.types';

const BaseTypography: FC<BaseTypographyProps> = ({
  as: Tag,
  children,
  ...props
}) => <Tag {...props}>{children}</Tag>;

const Typography: FC<TypographyProps> = ({
  className,
  variant,
  style,
  align,
  children,
  ...props
}) => {
  const isValidVariant = variant && TYPOGRAPHY_VARIANTS[variant];

  return (
    <BaseTypography
      as={
        isValidVariant
          ? (TYPOGRAPHY_VARIANTS[variant].variantName as ElementType)
          : 'p'
      }
      className={classNames(
        SHARED_CLASSNAMES,
        isValidVariant
          ? TYPOGRAPHY_VARIANTS[variant].variantClassName
          : TYPOGRAPHY_VARIANTS.body1.variantClassName,
        className
      )}
      style={{ textAlign: align, ...style }}
      {...props}
    >
      {children}
    </BaseTypography>
  );
};

export { Typography };
