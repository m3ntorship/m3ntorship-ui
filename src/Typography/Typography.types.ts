import { ElementType, HTMLAttributes, ReactNode } from 'react';

export const SHARED_CLASSNAMES = 'font-inter text-dgrey-800';
export const TYPOGRAPHY_VARIANTS = {
  hero: {
    variantName: 'h1',
    variantClassName: 'font-bold leading-80 text-72 tracking-very-tight',
  },
  h1: {
    variantName: 'h1',
    variantClassName: 'font-bold leading-72 text-64 tracking-tight',
  },
  h2: {
    variantName: 'h2',
    variantClassName: 'font-bold leading-64 text-56 tracking-semi-tight3',
  },
  h3: {
    variantName: 'h3',
    variantClassName: 'font-bold leading-56 text-48 tracking-semi-tight3',
  },
  h4: {
    variantName: 'h4',
    variantClassName: 'font-bold leading-48 text-40 tracking-semi-tight2',
  },
  h5: {
    variantName: 'h5',
    variantClassName: 'font-bold leading-40 text-32 tracking-semi-tight1',
  },
  h6: {
    variantName: 'h6',
    variantClassName: 'font-semibold leading-36 text-28 tracking-normal',
  },
  subtitle: {
    variantName: 'h6',
    variantClassName: 'font-regular leading-32 text-20 tracking-semi-wide',
  },
  body1: {
    variantName: 'p',
    variantClassName: 'font-regular leading-24 text-base tracking-semi-wide',
  },
  body2: {
    variantName: 'p',
    variantClassName: 'font-regular leading-20 text-14 tracking-semi-wide',
  },
  caption: {
    variantName: 'p',
    variantClassName: 'font-regular leading-16 text-12 tracking-wide',
  },
  disclaimer: {
    variantName: 'p',
    variantClassName: 'font-regular leading-14 text-10 tracking-very-wide',
  },
};

export interface TypographyProps
  extends HTMLAttributes<HTMLHeadingElement>,
    HTMLAttributes<HTMLParagraphElement> {
  variant?: keyof typeof TYPOGRAPHY_VARIANTS;
  children?: ReactNode;
}

export interface BaseTypographyProps
  extends HTMLAttributes<HTMLHeadingElement>,
    HTMLAttributes<HTMLParagraphElement> {
  as: ElementType;
}
