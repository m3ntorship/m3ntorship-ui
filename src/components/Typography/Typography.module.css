.hero {
  @apply font-bold leading-80 text-72 tracking-very-tight;
}
.h1 {
  @apply font-bold leading-72 text-64 tracking-tight;
}
.h2 {
  @apply font-bold leading-64 text-56 tracking-semi-tight3;
}
.h3 {
  @apply font-bold leading-56 text-48 tracking-semi-tight3;
}
.h4 {
  @apply font-bold leading-48 text-40 tracking-semi-tight2;
}
.h5 {
  @apply font-bold leading-40 text-32 tracking-semi-tight1;
}
.h6 {
  @apply font-semi-bold leading-36 text-28 tracking-normal;
}
.subtitle {
  @apply font-regular leading-32 text-20 tracking-semi-wide;
}
.body1 {
  @apply font-regular leading-24 text-base tracking-semi-wide;
}
.body2 {
  @apply font-regular leading-20 text-14 tracking-semi-wide;
}
.caption {
  @apply font-regular leading-16 text-12 tracking-wide;
}
.disclaimer {
  @apply font-regular leading-14 text-10 tracking-very-wide;
}
