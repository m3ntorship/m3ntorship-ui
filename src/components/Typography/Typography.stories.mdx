import { Meta, Story, Canvas } from '@storybook/addon-docs';
import { Typography } from './index';

<Meta title="Typography" component={Typography} />

export const Template = args => <Typography {...args} />;

# Typography

Typography component makes it easier to use default set of typography styles.

## Variants

Users can change the style of the typography component by passing `variant` prop.

<Canvas>
  <Story name="Hero" args={{variant: "hero", children: "Hero: Almost before we knew it, we had left the ground."}}>
    {Template.bind({})}
  </Story>
  <Story name="H1" args={{variant: "h1", children: "H1: Almost before we knew it, we had left the ground."}}>
    {Template.bind({})}
  </Story>
  <Story name="H2" args={{variant: "h2", children: "H2: Almost before we knew it, we had left the ground."}}>
    {Template.bind({})}
  </Story>
  <Story name="H3" args={{variant: "h3", children: "H3: Almost before we knew it, we had left the ground."}}>
    {Template.bind({})}
  </Story>
  <Story name="H4" args={{variant: "h4", children: "H4: Almost before we knew it, we had left the ground."}}>
    {Template.bind({})}
  </Story>
  <Story name="H5" args={{variant: "h5", children: "H5: Almost before we knew it, we had left the ground."}}>
    {Template.bind({})}
  </Story>
<Story name="H6" args={{variant: "h6", children: "H6: Almost before we knew it, we had left the ground."}}>
    {Template.bind({})}
  </Story>
</Canvas>

## Semantic Element

Users can change the semantic element of the typography component by passing `component` prop.

```jsx
// This will be rendered as `p`
<Typography component="p">
  Paragraph: Almost before we knew it, we had left the ground.
</Typography>

// This will be rendered as `h1`
<Typography component="h1">
  Paragraph: Almost before we knew it, we had left the ground.
</Typography>
```
