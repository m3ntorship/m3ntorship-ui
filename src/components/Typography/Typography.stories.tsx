import { Meta, Story } from '@storybook/react';
import React from 'react';

import { Typography } from '.';
import { TypographyProps } from './Typography.types';

export default {
  title: 'Typography',
  component: Typography,
  argTypes: {
    component: {
      control: { type: 'text' },
    },
  },
} as Meta<typeof Typography>;

const Template: Story<TypographyProps> = args => <Typography {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: 'Lorem ipsum dolor sit amet.',
};

export const WithVariant = Template.bind({});
WithVariant.args = {
  children: 'Lorem ipsum dolor sit amet.',
  variant: 'hero',
};

export const WithClassName = Template.bind({});
WithClassName.args = {
  children: 'Lorem ipsum dolor sit amet.',
  className: 'bg-red-200',
};

export const WithAlignment = Template.bind({});
WithAlignment.args = {
  children: 'Lorem ipsum dolor sit amet.',
  align: 'center',
};

export const WithComponent = Template.bind({});
WithComponent.args = {
  children: 'Lorem ipsum dolor sit amet.',
  component: 'div',
};
