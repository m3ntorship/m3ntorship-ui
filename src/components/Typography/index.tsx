import classNames from 'classnames';
import React, { FC } from 'react';

import styles from './Typography.module.css';

import { TypographyProps } from './Typography.types';

/** The Typography component makes it easy to apply a default set of font weights and sizes in the application. */
const Typography: FC<TypographyProps> = ({
  className,
  variant = 'body1',
  align = 'left',
  children,
  component,
  ...props
}) => {
  const Tag = component || 'p';
  return (
    <Tag
      className={classNames(
        'font-inter text-dgrey-800',
        `text-${align}`, //we added the text alignment classes to "safelist" key in the tailwind config file to make sure that Tailwind generates them.
        styles[variant],
        className
      )}
      {...props}
    >
      {children}
    </Tag>
  );
};

export { Typography };
