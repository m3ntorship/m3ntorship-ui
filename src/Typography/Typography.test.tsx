import React from 'react';
import * as renderer from 'react-test-renderer';
import { Typography } from './Typography';

describe('Typography', () => {
  it('Should render h1', () => {
    const tree = renderer.create(
      <Typography variant="h1">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render h2', () => {
    const tree = renderer.create(
      <Typography variant="h2">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render h3', () => {
    const tree = renderer.create(
      <Typography variant="h3">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render h4', () => {
    const tree = renderer.create(
      <Typography variant="h4">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render h5', () => {
    const tree = renderer.create(
      <Typography variant="h5">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render h6', () => {
    const tree = renderer.create(
      <Typography variant="h6">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render hero', () => {
    const tree = renderer.create(
      <Typography variant="hero">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render body1', () => {
    const tree = renderer.create(
      <Typography variant="body1">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render body2', () => {
    const tree = renderer.create(
      <Typography variant="body2">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render subtitle', () => {
    const tree = renderer.create(
      <Typography variant="subtitle">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render caption', () => {
    const tree = renderer.create(
      <Typography variant="caption">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
  it('Should render disclaimer', () => {
    const tree = renderer.create(
      <Typography variant="disclaimer">Lorem ipsum dolor sit amet.</Typography>
    );
    expect(tree).toMatchSnapshot();
  });
});
