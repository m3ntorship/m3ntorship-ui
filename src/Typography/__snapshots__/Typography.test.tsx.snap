// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography Should render body1 1`] = `
<p
  className="font-inter text-dgrey-800 font-regular leading-24 text-base tracking-semi-wide"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</p>
`;

exports[`Typography Should render body2 1`] = `
<p
  className="font-inter text-dgrey-800 font-regular leading-20 text-14 tracking-semi-wide"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</p>
`;

exports[`Typography Should render caption 1`] = `
<p
  className="font-inter text-dgrey-800 font-regular leading-16 text-12 tracking-wide"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</p>
`;

exports[`Typography Should render disclaimer 1`] = `
<p
  className="font-inter text-dgrey-800 font-regular leading-14 text-10 tracking-very-wide"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</p>
`;

exports[`Typography Should render h1 1`] = `
<h1
  className="font-inter text-dgrey-800 font-bold leading-72 text-64 tracking-tight"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</h1>
`;

exports[`Typography Should render h2 1`] = `
<h2
  className="font-inter text-dgrey-800 font-bold leading-64 text-56 tracking-semi-tight3"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</h2>
`;

exports[`Typography Should render h3 1`] = `
<h3
  className="font-inter text-dgrey-800 font-bold leading-56 text-48 tracking-semi-tight3"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</h3>
`;

exports[`Typography Should render h4 1`] = `
<h4
  className="font-inter text-dgrey-800 font-bold leading-48 text-40 tracking-semi-tight2"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</h4>
`;

exports[`Typography Should render h5 1`] = `
<h5
  className="font-inter text-dgrey-800 font-bold leading-40 text-32 tracking-semi-tight1"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</h5>
`;

exports[`Typography Should render h6 1`] = `
<h6
  className="font-inter text-dgrey-800 font-semibold leading-36 text-28 tracking-normal"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</h6>
`;

exports[`Typography Should render hero 1`] = `
<h1
  className="font-inter text-dgrey-800 font-bold leading-80 text-72 tracking-very-tight"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</h1>
`;

exports[`Typography Should render subtitle 1`] = `
<h6
  className="font-inter text-dgrey-800 font-regular leading-32 text-20 tracking-semi-wide"
  style={
    Object {
      "textAlign": undefined,
    }
  }
>
  Lorem ipsum dolor sit amet.
</h6>
`;
